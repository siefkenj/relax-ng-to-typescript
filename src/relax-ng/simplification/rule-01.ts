import { Plugin } from "unified";
import { remove } from "unist-util-remove";
import type { Test } from "unist-util-is";
import type { Root } from "xast";
import { NG_XML_NODES } from "../types";

/**
 * Simplification steps from https://relaxng.org/spec-20011203.html
 */

/**
 * Filter out all foreign nodes/elements. We don't filter foreign attributes, even though this is part of the spec.
 */
export const rule1: Plugin<void[], Root, Root> = function () {
    function shouldRemove(node: Root["children"][number] | Root): boolean {
        if (node.type === "root") {
            return false;
        }
        if (node.type !== "element" && node.type !== "text") {
            return true;
        }
        if (node.type === "text") {
            return false;
        }
        if (NG_XML_NODES.has(node.name)) {
            return false;
        }
        return true;
    }

    return (tree) => {
        remove(tree, shouldRemove as Test);
    };
};
