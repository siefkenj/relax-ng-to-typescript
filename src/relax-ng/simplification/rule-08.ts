import { Plugin } from "unified";
import { visit } from "unist-util-visit";
import { Root } from "xast";
import { isElement } from "../../xast-utils";

function isNotWhitespace(c: Root["children"][number]): boolean {
    if (c.type === "text" && c.value.trim() === "") {
        return false;
    }
    return true;
}

/**
 * Simplification steps from https://relaxng.org/spec-20011203.html
 */

/**
  The name attribute on an element or attribute element is transformed into a name child element.
 */
export const rule8: Plugin<void[], Root, Root> = function () {
    return (tree) => {
        tree.children = tree.children.filter(isNotWhitespace);
        visit(tree, isElement, (node) => {
            if (node.name !== "element" && node.name !== "attribute") {
                return;
            }
            if (node.attributes?.name) {
                const name = node.attributes.name;
                delete node.attributes.name;
                node.children.unshift({
                    type: "element",
                    name: "name",
                    children: [{ type: "text", value: name }],
                    attributes: {},
                });
            }
        });
    };
};
