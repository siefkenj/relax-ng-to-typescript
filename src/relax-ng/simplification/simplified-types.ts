/**
 * RELAX-NG types for the schema after simplification. Ported from https://relaxng.org/spec-20011203.html#element-pattern
 */

import { Element as XMLElement, Text as XMLText, Root as XMLRoot } from "xast";

export interface NGSimpRoot extends Omit<XMLRoot, "children"> {
    children: [NGSimpGrammar];
}

export interface Element<Name extends string> extends XMLElement {
    name: Name;
    attributes: never;
    children: never[];
}

export interface ElementWithChildren<
    Name extends string,
    Children extends XMLElement["children"]
> extends XMLElement {
    name: Name;
    attributes: never;
    children: Children;
}

export interface NGSimpGrammar extends Omit<XMLElement, "attributes"> {
    name: "grammar";
    //attributes?: never;
    children: [NGSimpStart, ...NGSimpDefine[]];
}

export interface NGSimpStart extends XMLElement {
    name: "start";
    attributes: never;
    children: [NGSimpTop];
}

export interface NGSimpDefine extends XMLElement {
    name: "define";
    attributes: { name: string };
    children: [NGSimpElement];
}
export interface NGSimpRef extends XMLElement {
    name: "ref";
    attributes: { name: string };
    children: never[];
}

export type NGSimpElement = ElementWithChildren<
    "element",
    [NGSimpNameClass, NGSimpTop]
>;

export type NGSimpNotAllowed = Element<"notAllowed">;
export type NGSimpEmpty = Element<"empty">;
export type NGSimpText = Element<"text">;

export type NGSimpTop = NGSimpNotAllowed | NGSimpPattern;
export type NGSimpPattern = NGSimpEmpty | NGSimpNonEmptyPattern;

export type NGSimpNonEmptyPattern =
    | NGSimpText
    | NGSimpData
    | NGSimpValue
    | ElementWithChildren<"list", [NGSimpPattern]>
    | ElementWithChildren<"attribute", [NGSimpNameClass, NGSimpPattern]>
    | NGSimpRef
    | ElementWithChildren<"oneOrMore", [NGSimpNonEmptyPattern]>
    | ElementWithChildren<"choice", [NGSimpPattern, NGSimpNonEmptyPattern]>
    | ElementWithChildren<
          "group",
          [NGSimpNonEmptyPattern, NGSimpNonEmptyPattern]
      >
    | ElementWithChildren<
          "interleave",
          [NGSimpNonEmptyPattern, NGSimpNonEmptyPattern]
      >;

export interface NGSimpData extends XMLElement {
    name: "data";
    attributes: { type: string };
    children: (NGSimpParam | NGSimpExceptPattern)[];
}

export interface NGSimpValue extends XMLElement {
    name: "value";
    attributes: { type: string };
    children: XMLText[];
}

export interface NGSimpParam extends XMLElement {
    name: "param";
    attributes: { name: string };
    children: XMLText[];
}
export interface NGSimpName extends XMLElement {
    name: "name";
    attributes: never;
    children: XMLText[];
}

export interface NGSimpExceptPattern extends XMLElement {
    name: "except";
    attributes: never;
    children: [NGSimpPattern];
}

export interface NGSimpExceptNameClass extends XMLElement {
    name: "except";
    attributes: never;
    children: NGSimpNameClass[];
}
export type NGSimpNameClass =
    | ElementWithChildren<"anyName", NGSimpExceptNameClass[]>
    | ElementWithChildren<"nsName", NGSimpExceptNameClass[]>
    | NGSimpName
    | ElementWithChildren<"choice", [NGSimpNameClass, NGSimpNameClass]>;
