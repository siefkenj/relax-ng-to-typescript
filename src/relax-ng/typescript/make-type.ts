import { visit } from "unist-util-visit";
import { toXml } from "xast-util-to-xml";
import { expected, isElement } from "../../xast-utils";
import {
    ElementTypeDescriptor,
    extractElementType,
} from "../extract/element-type";
import {
    NGSimpDefine,
    NGSimpGrammar,
    NGSimpRef,
} from "../simplification/simplified-types";
import { normalizeTypeName } from "./normalize-type-name";
import { NGElement, NGMethod } from "../types";

// Basic XML element names
const XML_ELEMENT = "XMLElement";
const XML_TEXT = "XMLText";
const XML_ELEMENT_WITH_NO_ATTRIBUTES = `Omit<${XML_ELEMENT}, "attributes">`;

const TYPES_PREAMBLE = `
/**
 * This file was AUTOGENERATED from a RELAX-NG grammar.
 */

import {Element as XMLElement, Text as XMLText} from "xast"\n\n`;

/**
 * Put quote marks around an XML attribute name if TypeScript would need quotes.
 */
function quoteAttributeIfNeeded(attr: string): string {
    if (attr.match(/[^a-zA-Z0-9_]/) || attr.charAt(0).match(/[0-9]/)) {
        return JSON.stringify(attr);
    }
    return attr;
}

/**
 * Outputs TypeScript code for a type representing the element described by `desc`. If
 * `refName` is supplied, it will be used as the type name. Otherwise the element name
 * will be used.
 */
export function makeElementType(desc: ElementTypeDescriptor, refName?: string) {
    const hasAttributes = Object.keys(desc.attributes).length > 0;
    const name = refName ? refName : normalizeTypeName(desc.name, "Element");

    const ret: string[] = [];

    if (hasAttributes) {
        ret.push(`interface ${name} extends ${XML_ELEMENT} {`);
    } else {
        ret.push(
            `interface ${name} extends ${XML_ELEMENT_WITH_NO_ATTRIBUTES} {`
        );
    }
    ret.push(`  name: ${JSON.stringify(desc.name)};`);
    if (hasAttributes) {
        ret.push(`  attributes: {`);
        ret.push(
            ...Object.entries(desc.attributes).map(
                ([attr, attrDesc]) =>
                    `    ${quoteAttributeIfNeeded(attr)}${
                        attrDesc.optional ? "?" : ""
                    }: ${attrDesc.type.join(" | ")};`
            )
        );
        ret.push("  };");
    }

    const children = desc.children.map((ref) => normalizeTypeName(ref.ref));
    if (desc.textChildrenAllowed) {
        children.push(XML_TEXT);
    }
    children.sort();

    if (children.length === 0) {
        ret.push("  children: never[];");
    } else {
        const typeStr =
            children.length === 1
                ? `${children[0]}[]`
                : `(${children.join(" | ")})[]`;
        ret.push(`  children: ${typeStr};`);
    }

    ret.push("}");

    const typeStr = ret.join("\n");
    return { typeStr, dependsOn: children };
}

export type JSONGrammar = {
    startType: string;
    refs: Record<string, JSONGrammarItem>;
};
export type JSONGrammarItem =
    | { type: "text" }
    | { type: "unknown" }
    | ReturnType<typeof extractElementType>;

/**
 * Output all types specified by the grammar in both TypeScript
 * form and JSON form.
 */
export function makeTypesForGrammar(grammar: NGSimpGrammar): {
    typescriptStr: string;
    grammar: JSONGrammar;
} {
    const start = grammar.children[0];
    const allDefs = grammar.children.slice(1) as NGSimpDefine[];
    let startRef = start.children[0];
    //console.log(start.children);
    try {
        expected(startRef, "ref");
    } catch {
        startRef = findElementRefInChoices(startRef as any, "ElementPretext");
        expected(startRef, "ref");
    }

    const allDefsMap: Record<string, NGSimpDefine> = Object.fromEntries(
        allDefs.map((def) => [def.attributes.name, def])
    );

    // `XMLText` is a special node that we create separately
    const exportedRefs: Record<string, JSONGrammarItem> = {
        XMLText: { type: "text" },
    };

    const interfaces: string[] = [
        `type StartElement = ${startRef.attributes.name};`,
    ];
    const queue: string[] = [startRef.attributes.name];
    // Recursively export all the types that are needed
    while (queue.length > 0) {
        const refName = queue.pop() || "";
        if (!refName || exportedRefs[refName]) {
            continue;
        }
        const ref = allDefsMap[refName];
        if (!ref) {
            // We encountered a type that doesn't have a definition.
            interfaces.push(generateTypeForMissingRef(refName));
            exportedRefs[refName] = { type: "unknown" };
            continue;
        }
        //if (refName.includes("Author")) {
        //    console.log("doing", refName);
        //}
        const elm = ref.children[0];
        const typeDesc = extractElementType(elm);
        const { typeStr, dependsOn } = makeElementType(typeDesc, refName);
        interfaces.push(typeStr);
        queue.push(...dependsOn);
        exportedRefs[refName] = typeDesc;

        //if (refName.includes("Author")) {
        //    console.log(typeStr);
        //}
    }

    return {
        typescriptStr:
            TYPES_PREAMBLE +
            "\n\n" +
            interfaces.map((i) => `export ${i}`).join("\n\n"),
        grammar: { startType: startRef.attributes.name, refs: exportedRefs },
    };
}

/**
 * Generate a type for a type named `typeName` that doesn't have a corresponding
 * `<define name="${typeName}">` tag in the grammar.
 */
function generateTypeForMissingRef(typeName: string): string {
    return `type ${typeName} = unknown;`;
}

/**
 * Drill down a `<choice>...</choice>` blocks and find the first ref
 * whose initial pattern matches the given `namePrefix`.
 */
function findElementRefInChoices(elm: NGMethod, namePrefix: string): NGSimpRef {
    const allFlatRefs: NGSimpRef[] = [];
    visit(elm as any, "element", (el) => {
        if (el.name === "ref") allFlatRefs.push(el);
    });

    const ret = allFlatRefs.find((ref) =>
        ref.attributes.name.startsWith(namePrefix)
    );

    if (ret) {
        return ret;
    }
    throw new Error(`Could not find ref with prefix ${namePrefix}`);
}
