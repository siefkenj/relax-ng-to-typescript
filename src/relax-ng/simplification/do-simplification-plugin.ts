import { Plugin } from "unified";
import type { Root } from "xast";
import { rule1 } from "./rule-01";
import { rule2 } from "./rule-02";
import { rule3 } from "./rule-03";
import { rule4 } from "./rule-04";
import { rule5 } from "./rule-05";
import { rule6 } from "./rule-06";
import { rule7 } from "./rule-07";
import { rule8 } from "./rule-08";
import { rule9 } from "./rule-09";
import { rule10 } from "./rule-10";
import { rule11 } from "./rule-11";
import { rule12 } from "./rule-12";
import { rule13 } from "./rule-13";
import { rule14 } from "./rule-14";
import { rule15 } from "./rule-15";
import { rule16 } from "./rule-16";
import { rule17 } from "./rule-17";
import { rule18 } from "./rule-18";
import { rule19 } from "./rule-19";
import { rule20 } from "./rule-20";
import { rule21 } from "./rule-21";
import { NGSimpRoot } from "./simplified-types";

/**
 * Apply simplification rules defined in https://relaxng.org/spec-20011203.html
 *
 * Note: Rules related to XML namespaces are skipped.
 */
export const doSimplificationPlugin: Plugin<never[], Root, NGSimpRoot> =
    function () {
        this.use(rule1)
            .use(rule2)
            .use(rule3)
            .use(rule4)
            .use(rule5)
            .use(rule6)
            .use(rule7)
            .use(rule8)
            .use(rule9)
            .use(rule10)
            .use(rule11)
            .use(rule12)
            .use(rule13)
            .use(rule14)
            .use(rule15)
            .use(rule16)
            .use(rule17)
            .use(rule18)
            .use(rule19)
            .use(rule20)
            .use(rule21);
    };
