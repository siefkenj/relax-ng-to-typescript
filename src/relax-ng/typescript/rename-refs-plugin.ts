import { Plugin } from "unified";
import { visit } from "unist-util-visit";
import { isElement } from "../../xast-utils";
import { NGSimpRoot } from "../simplification/simplified-types";
import { normalizeTypeName } from "./normalize-type-name";
import { Root, Element as XMLElement, Text as XMLText } from "xast";

/**
 * UnifiedJs plugin that renames all references to make sure they start with the word `Element`
 * and that they are valid TypeScript type names.
 */
export const renameRefsPlugin: Plugin<never[], NGSimpRoot, NGSimpRoot> =
    function () {
        return (tree) => {
            // Compute what all the existing ref names are and what they should be renamed to
            const refNames: Record<string, string> = {};
            visit(tree, isElement, (node: XMLElement) => {
                if (!(node.name === "define")) {
                    return;
                }
                refNames[node.attributes.name!] = node.attributes.name!;
            });
            // We iterate over keys instead of doing a `for in` loop because we mutate
            // `refNames` in the loop.
            for (const name of Object.keys(refNames)) {
                if (!name.startsWith("Element")) {
                    let newName = normalizeTypeName(name, "Element");
                    if (refNames[newName]) {
                        // We need to search for an unused name
                        let i = 2;
                        while (refNames[`${newName}${i}`] && i < 1000) {
                            i++;
                        }
                        if (i === 1000) {
                            console.warn(
                                `Tried to make a unique version of the name "${newName}" but failed after ${i} attempts`
                            );
                        }
                        newName = `${newName}${i}`;
                    }
                    refNames[name] = newName;
                    refNames[newName] = newName;
                }
            }

            // First thing to do is to rename all refs as their Typescript type names.
            const usedNames: Set<string> = new Set();
            visit(tree, isElement, (node: XMLElement) => {
                if (node.name === "ref") {
                    if (refNames[node.attributes.name!]) {
                        node.attributes.name = refNames[node.attributes.name!];
                    } else {
                        const newName = normalizeTypeName(
                            node.attributes.name!,
                            "Element"
                        );
                        console.warn(
                            `Found <ref name="${node.attributes.name}" /> reference, but did not find a corresponding <define>; renaming to "${newName}"`
                        );
                        node.attributes.name = newName;
                    }
                }
                if (node.name === "define") {
                    node.attributes.name = refNames[node.attributes.name!];
                }
            });
            return tree;
        };
    };
