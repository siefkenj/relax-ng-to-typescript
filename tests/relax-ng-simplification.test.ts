import { describe, it, expect } from "vitest";
import util from "node:util";
import Prettier from "prettier";
import { unifiedXml } from "./utils";
import {
    rule1,
    rule12,
    rule2,
    rule4,
    rule8,
} from "../src/relax-ng/simplification";
import { doSimplificationPlugin } from "../src/relax-ng/simplification/do-simplification-plugin";
import { renameRefsPlugin } from "../src/relax-ng/typescript/rename-refs-plugin";

// Make console.log pretty-print by default
const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

describe("relax-ng-simplify", () => {
    it("rule1", async () => {
        const processor = unifiedXml().use(rule1);
        const source = `<grammar>
    <element name="foo">
        <a:documentation>Important prop!</a:documentation>
        <text />
    </element>
</grammar>`;
        const processed = String(processor.processSync(source));

        const formatted = Prettier.format(processed, { parser: "html" });
        expect(formatted).toMatchSnapshot();
    });

    it("rule2", async () => {
        const processor = unifiedXml().use(rule1).use(rule2);
        const source = `<grammar>
    <element name="foo">
        <a:documentation>Important prop!</a:documentation>
        <text />
    </element>
</grammar>`;
        const processed = String(processor.processSync(source));

        const formatted = Prettier.format(processed, { parser: "html" });
        expect(formatted).toMatchSnapshot();
    });

    it("rule4", async () => {
        const processor = unifiedXml().use(rule1).use(rule2).use(rule4);
        const source = `<grammar>
    <element name="foo">
        <attribute name="attr"><choice><value>7</value><value type="integer">9</value></choice></attribute>
        <a:documentation>Important prop!</a:documentation>
        <text />
    </element>
</grammar>`;
        const processed = String(processor.processSync(source));

        const formatted = Prettier.format(processed, { parser: "html" });
        expect(formatted).toMatchSnapshot();
    });

    it("rule4", async () => {
        const processor = unifiedXml()
            .use(rule1)
            .use(rule2)
            .use(rule4)
            .use(rule8);
        const source = `<grammar>
    <element name="foo">
        <attribute name="attr"><choice><value>7</value><value type="integer">9</value></choice></attribute>
        <a:documentation>Important prop!</a:documentation>
        <text />
    </element>
</grammar>`;
        const processed = String(processor.processSync(source));

        const formatted = Prettier.format(processed, { parser: "html" });
        expect(formatted).toMatchSnapshot();
    });

    it("rule4", async () => {
        const processor = unifiedXml()
            .use(rule1)
            .use(rule2)
            .use(rule8)
            .use(rule12);
        const source = `
<element name="ol">
  <optional>
    <attribute name="cols"/>
  </optional>
  <optional>
    <attribute name="marker"/>
  </optional>
  <oneOrMore>
      <ref name="li"/>
  </oneOrMore>
</element>
`;
        const processed = String(processor.processSync(source));

        const formatted = Prettier.format(processed, { parser: "html" });
        expect(formatted).toMatchSnapshot();
    });
});
