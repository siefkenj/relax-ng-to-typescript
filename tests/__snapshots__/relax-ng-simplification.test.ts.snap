// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`relax-ng-simplify > rule1 1`] = `
"<grammar>
  <element name=\\"foo\\">
    <text></text>
  </element>
</grammar>
"
`;

exports[`relax-ng-simplify > rule2 1`] = `
"<grammar
  ><element name=\\"foo\\"><text></text></element
></grammar>
"
`;

exports[`relax-ng-simplify > rule4 1`] = `
"<grammar
  ><element name=\\"foo\\"
    ><attribute name=\\"attr\\"
      ><choice
        ><value type=\\"token\\">7</value><value type=\\"integer\\">9</value></choice
      ></attribute
    ><text></text></element
></grammar>
"
`;

exports[`relax-ng-simplify > rule4 2`] = `
"<grammar
  ><element
    ><name>foo</name
    ><attribute
      ><name>attr</name
      ><choice
        ><value type=\\"token\\">7</value><value type=\\"integer\\">9</value></choice
      ></attribute
    ><text></text></element
></grammar>
"
`;

exports[`relax-ng-simplify > rule4 3`] = `
"<element
  ><name>ol</name
  ><group
    ><group
      ><optional
        ><attribute><name>cols</name><text></text></attribute></optional
      ><optional
        ><attribute
          ><name>marker</name><text></text></attribute></optional></group
    ><oneOrMore><ref name=\\"li\\"></ref></oneOrMore></group
></element>
"
`;
